packages/docusaurus-theme-common/src/utils/useChangeRoute.ts

```diff
@@ -9,11 +9,11 @@ import {useRef, useEffect} from 'react';
 import {useLocation} from '@docusaurus/router';
 
 export function useChangeRoute(onRouteChange: () => void): void {
-  const {pathname} = useLocation();
+  const {pathname, hash} = useLocation();
   const latestPathnameRef = useRef(pathname);
 
   useEffect(() => {
-    if (pathname !== latestPathnameRef.current) {
+    if (!hash && pathname !== latestPathnameRef.current) {
       latestPathnameRef.current = pathname;
       onRouteChange();
```

Comment:
What about using `onRouteChange({location})` instead, and preventing the call to `programmaticFocus` if there's a hash?

`onRouteChange` should rather fire even if an anchor link is clicked, for example this should be triggered, otherwise the hook does not really do what its name implies.

We also may want this callback to keep being called for anchor links:

```js
  useChangeRoute(() => {
    if (!hideOnScroll) {
      return;
    }

    setIsNavbarVisible(true);
  });
```

Reply:
I thought about this at first, but for the hideable navbar to work properly we also need to ignore hash changes in the according hook, otherwise the navbar will automatically show up when go to on anchor link on cross-page. So in order not to duplicate code, we can slightly improve `useChangedRoute` hook directly, because at the moment it also triggers when hash has changed, but it shouldn't. After all, this hook is only supposed to keep track of changes in current pathname, right?

Reply:
Yes but it won't trigger the callback when current pathname change, if a hash is also present, which is confusing.

Agree to merge that for now, but not 100% sure it'll be the best long-term decision. It might become a problem if we decide to use this hook for another use-case. For example it wouldn't work reliably anymore for handling analytics or things like that, because it wouldn't report page changes if user keeps clicking on anchor links.

Reply:
Actually I'll do that now, because I'm not comfortable at all having the `latestPathnameRef` not being updated when there is a hash. This might lead to subtle bugs in this hook after the user has navigated to an anchor link

Reply:
Alright, although there is more code/complexity, agree, it looks more flexible and reliable :+1: